src/fsharp/block.fs

```diff
@@ -98,85 +96,55 @@ module Block =
             let builder = ImmutableArray.CreateBuilder(n)
             for i = 0 to n - 1 do
                 builder.Add(mapper i arr1.[i] arr2.[i])
-            builder.ToImmutable()
-
-    let concat (arr: block<block<'T>>) : block<'T> =
-        if arr.IsEmpty then
-            empty
-        elif arr.Length = 1 then
-            arr.[0]
-        elif arr.Length = 2 then
-            arr.[0].AddRange(arr.[1])
-        else
-            let mutable arr2 = arr.[0]
-            for i = 1 to arr.Length - 1 do
-                arr2 <- arr2.AddRange(arr.[i])
-            arr2
+            builder.MoveToImmutable()
+
+    let concat (arrs: block<block<'T>>) : block<'T> =
+        match arrs.Length with
+        | 0 -> ImmutableArray.Empty
+        | 1 -> arrs.[0]
+        | 2 -> arrs.[0].AddRange(arrs.[1])
+        | _ ->
+            let mutable acc = 0    
+            for h in arrs do
+                acc <- acc + h.Length
+
+            let builder = ImmutableArray.CreateBuilder(acc)
+            for i = 0 to arrs.Length - 1 do
+                builder.AddRange(arrs.[i])
+            builder.MoveToImmutable()
```

Comment:
Missing capacity <- count or ToImmutable here, as capacity could be anything.

Reply:
I think this is fine. `acc` is the total length of the final array.

Reply:
Ah I completely missed the the length summing loop, yep it's fine!